### ISSUE TITLE
[Feature Request] Stay on screensaver / Navigate back to screensaver

### ISSUE THREAD

**Issue by bkbartk** [joBr99/nspanel-lovelace-ui#55]
Hello,

first of all, I really like the implementation you have created. It's really easy to configure and works really great.
I'm still strugling a little bit with the "CustomSend" commands. But I will hopefully find out how to manage them.

For now I have a few questions if this is possible of not.
- is it possible to also use the screensaver as some kind of homepage?
When I tab now to see the weather forecast the interface goes directly to the first page. But I like to see what's on the screensaver screen. I also like to be able to navigate back to it. So maybe some way to set this page as first page 
- is it possible to swipe left and right instead of using the arrows?
- is it possible to create some kind of overview page? If I have like 8 pages, it would be beneficial to have some kind of navigation.

With kind regards

Bart

**Comment by joBr99** [comment]
Hey,

for the swipe feature, that's something I've tried and it didn't work good enough. 
The screen in nspanel doesn't support swipe out of the box and you have to build hacky workarounds to get swipe support and they are limited, so you will end up with some swipe support, but you have to exclude all regions from the swipe that contain buttons or other touch controls. And this isn't a great experience, so swipe isn't possible.

For the other two points definitly possible, I will leave this open as a feature request, but it requires a redesign of the page configuration and the event flow around page/screensaver navigation, so this is a big chunk of work.

**Comment by bkbartk** [comment]
it's unfortunate that swipe isn't possible, still I think the design you have created is really cool.
thanks for accepting the other 2 as feature request.

**Comment by joBr99** [comment]
@bkbartk overview page feature isn't finished but already included in the newest release

```yaml
    pages:
      - type: cardGrid
        heading: Overview
        items:
          - cover.rolladenfenster_cover_1
          - type: cardEntities
            heading: Test Entities 1
            icon: lightbulb
            items:
              - switch.test_item
              - switch.test_item
              - navigate.cardEntities_Overview
              - delete
              
```

**Comment by bkbartk** [comment]
thnx,
I get the idea, at the moment my nested page stays empty, But I think this is because it isn't fully finished.

**Comment by joBr99** [comment]
hmm, this should already working, can you check the appdaemon log for any warnings?

**Comment by bkbartk** [comment]
yes,
I have an error, masked as warning.
> 2022-03-26 20:30:02.601763 INFO nspanel-kamer: Started rendering of page 1 with type cardEntities
2022-03-26 20:30:02.610272 INFO nspanel-kamer: Sending MQTT Message: pageType,cardEntities
2022-03-26 20:30:02.626944 INFO nspanel-kamer: Sending MQTT Message: entityUpdHeading,Test Entities 1
2022-03-26 20:30:02.660144 WARNING nspanel-kamer: ------------------------------------------------------------
2022-03-26 20:30:02.661750 WARNING nspanel-kamer: Unexpected error in worker for App nspanel-kamer:
2022-03-26 20:30:02.663505 WARNING nspanel-kamer: Worker Ags: {'id': '3c4d5e09a0c24f9c92d83c82d88e6340', 'name': 'nspanel-kamer', 'objectid': '355de119466c4602a0cf6c35b008b0dc', 'type': 'event', 'event': 'MQTT_MESSAGE', 'function': <bound method LuiMqttListener.mqtt_event_callback of <luibackend.mqttListener.LuiMqttListener object at 0x7fa8221af0>>, 'data': {'topic': 'tele/tasmota_NSPanel_kamer/RESULT', 'wildcard': None, 'payload': '{"CustomRecv":"event,buttonPress2,navigate.cardEntities_Test_Entities_1,button"}'}, 'pin_app': True, 'pin_thread': 0, 'kwargs': {'topic': 'tele/tasmota_NSPanel_kamer/RESULT', '__thread_id': 'thread-0'}}
2022-03-26 20:30:02.665251 WARNING nspanel-kamer: ------------------------------------------------------------
2022-03-26 20:30:02.668018 WARNING nspanel-kamer: Traceback (most recent call last):
  File "/usr/lib/python3.9/site-packages/appdaemon/threading.py", line 950, in worker
    funcref(args["event"], data, self.AD.events.sanitize_event_kwargs(app, args["kwargs"]))
  File "/config/appdaemon/apps/nspanel-lovelace-ui/luibackend/mqttListener.py", line 51, in mqtt_event_callback
    self._controller.button_press(entity_id, btype, value)
  File "/config/appdaemon/apps/nspanel-lovelace-ui/luibackend/controller.py", line 166, in button_press
    self._pages_gen.render_page(self._current_page)
  File "/config/appdaemon/apps/nspanel-lovelace-ui/luibackend/pages.py", line 273, in render_page
    self.generate_entities_page(heading, page.get_items())
  File "/config/appdaemon/apps/nspanel-lovelace-ui/luibackend/pages.py", line 186, in generate_entities_page
    command += self.generate_entities_item(item)
  File "/config/appdaemon/apps/nspanel-lovelace-ui/luibackend/pages.py", line 141, in generate_entities_item
    name = self._config.get_root_page().search_page_by_name(item)[0].data.get("heading")
IndexError: list index out of range
2022-03-26 20:30:02.669314 WARNING nspanel-kamer: ------------------------------------------------------------
2022-03-26 20:30:04.582694 INFO nspanel-kamer: Received Message from Screen: event,buttonPress2,cardEntities,bPrev
2022-03-26 20:30:04.590583 INFO nspanel-kamer: Button Press Event; entity_id: cardEntities; button_type: bPrev; value: None 
2022-03-26 20:30:04.592772 WARNING nspanel-kamer: ------------------------------------------------------------
2022-03-26 20:30:04.594305 WARNING nspanel-kamer: Unexpected error in worker for App nspanel-kamer:
2022-03-26 20:30:04.595849 WARNING nspanel-kamer: Worker Ags: {'id': '3c4d5e09a0c24f9c92d83c82d88e6340', 'name': 'nspanel-kamer', 'objectid': '355de119466c4602a0cf6c35b008b0dc', 'type': 'event', 'event': 'MQTT_MESSAGE', 'function': <bound method LuiMqttListener.mqtt_event_callback of <luibackend.mqttListener.LuiMqttListener object at 0x7fa8221af0>>, 'data': {'topic': 'tele/tasmota_NSPanel_kamer/RESULT', 'wildcard': None, 'payload': '{"CustomRecv":"event,buttonPress2,cardEntities,bPrev"}'}, 'pin_app': True, 'pin_thread': 0, 'kwargs': {'topic': 'tele/tasmota_NSPanel_kamer/RESULT', '__thread_id': 'thread-0'}}
2022-03-26 20:30:04.597122 WARNING nspanel-kamer: ------------------------------------------------------------
2022-03-26 20:30:04.598991 WARNING nspanel-kamer: Traceback (most recent call last):
  File "/usr/lib/python3.9/site-packages/appdaemon/threading.py", line 950, in worker
    funcref(args["event"], data, self.AD.events.sanitize_event_kwargs(app, args["kwargs"]))
  File "/config/appdaemon/apps/nspanel-lovelace-ui/luibackend/mqttListener.py", line 51, in mqtt_event_callback
    self._controller.button_press(entity_id, btype, value)
  File "/config/appdaemon/apps/nspanel-lovelace-ui/luibackend/controller.py", line 139, in button_press
    self._pages_gen.render_page(self._current_page)
  File "/config/appdaemon/apps/nspanel-lovelace-ui/luibackend/pages.py", line 266, in render_page
    page_type = config["type"]
TypeError: string indices must be integers
2022-03-26 20:30:04.600244 WARNING nspanel-kamer: ------------------------------------------------------------
2022-03-26 20:30:20.873891 INFO nspanel-kamer: Received Message from Screen: event,screensaverOpen
2022-03-26 20:30:20.941866 INFO nspanel-kamer: Sending MQTT Message: weatherUpdate,?23?11.9��C?26?72 %?zo?23?14.9 ��C?ma?23?15.8 ��C

**Comment by joBr99** [comment]
okay that's something I can improve what do you have configured within the nested page?

show me your config, this line is causeing the error

- navigate.cardEntities_Overview

**Comment by bkbartk** [comment]
btw,
If I change 
`- navigate.cardEntities_Overview`
to 
`- delete`
the page shows up, I only have no way to navigate back

I plain copied your sample config to test

**Comment by joBr99** [comment]
yes, currently this line is your way back

needs to be navigate.type_heading of one of your other pages "." "." and " " needs to be replaced with "_"

**Comment by joBr99** [comment]
yep I copy pasted this from my test config and changed only the outer page to grid, because that's more like an overview page :)

**Comment by bkbartk** [comment]
the cool thing is, that without nesting I still can use navigate

>       - type: cardGrid
        heading: Overview
        items:
          - navigate.cardEntities_Home
          - navigate.cardGrid_Keuken
          - navigate.cardGrid_Boven
          - navigate.cardThermo_Thermostat

**Comment by joBr99** [comment]
yep, I know :) however they still have to be defined somewhere in the tree

**Comment by bkbartk** [comment]
yes, but making a navigation without a page is a little bit strange to me.
my pages now looks like this(items collapsed)
![image](https://user-images.githubusercontent.com/17070805/160255034-8756bc10-4a92-4a5b-903c-704f837f02f1.png)

**Comment by joBr99** [comment]
so, this error should be fixed, in case there's no page found for the navigate config entry, it will now render "page not found"

navigate searchs for the page from the root page, so it's indented that you can jump to any page with it

**Comment by joBr99** [comment]
@bkbartk 

just added an config option for the screensaver to require 2 taps to wake, first will change the brightness of the screen, second will render the pages

```yaml
    doubleTapToUnlock: True
```

https://github.com/joBr99/nspanel-lovelace-ui#how-to-upgrade-from-a-release-to-the-current-development-version

**Comment by joBr99** [comment]
Hmm, I guess you have to update the firmware looks like the value is missing.

**Comment by bkbartk** [comment]
thnx,

looks great, 
when I tab only once, the screen doesn't get dark again. I need to tab a 2nd time for the screen to go dark again.
also would if be possible to create a button to go back to the screensaver? so I don't have to wain x(15) sec to see the weather forecast

**Comment by joBr99** [comment]
this needs some changes on the serial protcol around the navigation, which are related to #46 #55 and #83, so this might take a while

**Comment by bkbartk** [comment]
No problem. I’m already really glad with what you have created.

**Comment by joBr99** [comment]
@bkbartk with the newest release the screensaver will get dark again

**Comment by joBr99** [comment]
with the newest release you can navigate back to the screensaver with a navigate item
```yaml
  config:
    screensaver:
      type: screensaver
      entity: weather.k3ll3r
      key: screensaver
      title: Screensaver
    cards:
      - type: cardEntities
        title: Test
        entities:
          - entity: navigate.screensaver_screensaver
```

**Comment by bkbartk** [comment]
sorry,

didn't have time to update the component earlier.
but when I use it like this. when I get back to the screensaver I get stuck,
there is no way to navigate away from it.

**Comment by joBr99** [comment]
yes, because navigate will set screensaver to the current_page and exiting screensaver will render the current page, I will fix this with #123

**Comment by joBr99** [comment]
@bkbartk with the recent changes for #123, this should be working now, I was able to navigate to the screensaver card aswell as navigating back

 (but I didn't do much testing with the navigation changes itself)
